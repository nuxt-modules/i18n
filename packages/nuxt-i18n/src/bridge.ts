import { resolve } from 'pathe'
import { resolveModule, addPluginTemplate } from '@nuxt/kit'
import { distDir } from './dirs'

import type { Nuxt } from '@nuxt/schema'

export async function setupNuxtBridge(nuxt: Nuxt) {
  // resolve vue-i18n as vue-i18n-legacy
  nuxt.options.alias['vue-i18n'] = resolveModule('vue-i18n-legacy/dist/vue-i18n.esm.js', {
    paths: nuxt.options.modulesDir
  })
  nuxt.options.build.transpile.push('vue-i18n')

  // resolve vue-i18n-bridge as vue-i18n
  const vueI18nBridgePath = nuxt.options.dev
    ? 'vue-i18n-bridge/dist/vue-i18n-bridge.esm-bundler.js'
    : 'vue-i18n-bridge/dist/vue-i18n-bridge.runtime.esm-bundler.js'
  nuxt.options.alias['vue-i18n-bridge'] = resolveModule(vueI18nBridgePath, {
    paths: nuxt.options.modulesDir
  })
  nuxt.options.build.transpile.push('vue-i18n-bridge')

  addPluginTemplate({
    filename: 'runtime/bridge.plugin.mjs',
    src: resolve(distDir, 'runtime/bridge.plugin.mjs')
  })
}
