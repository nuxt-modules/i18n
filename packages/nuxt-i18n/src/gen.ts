import createDebug from 'debug'
import { isString, isRegExp, isFunction, isArray, isObject } from '@intlify/shared'
import { templateUtils } from '@nuxt/kit'
import { genImport } from 'knitwork'

import type { NuxtI18nOptions, NuxtI18nInternalOptions, LocaleInfo, NoNullable } from './types'
import type { NuxtI18nOptionsDefault } from './constants'

export type LoaderOptions = {
  localeCodes?: string[]
  localeInfo?: LocaleInfo[]
  nuxtI18nOptions?: NuxtI18nOptions
  nuxtI18nOptionsDefault?: NuxtI18nOptionsDefault
  nuxtI18nInternalOptions?: NuxtI18nInternalOptions
}

const debug = createDebug('@nuxtjs/i18n:gen')

export function generateLoaderOptions(
  lazy: NoNullable<NuxtI18nOptions['lazy']>,
  langDir: NuxtI18nOptions['langDir'],
  options: LoaderOptions = {}
) {
  // TODO: lazy loading local info
  let genCode = ''
  const localeInfo = options.localeInfo || []
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const { fallbackLocale } = (isObject(options.nuxtI18nOptions?.vueI18n) ? options.nuxtI18nOptions?.vueI18n : {}) as any
  const syncLocaleFiles = new Set<LocaleInfo>()
  const asyncLocaleFiles = new Set<LocaleInfo>()

  if (langDir) {
    if (fallbackLocale && isString(fallbackLocale)) {
      const localeObject = localeInfo.find(l => l.code === fallbackLocale)
      if (localeObject) {
        syncLocaleFiles.add(localeObject)
      }
    }
    for (const locale of localeInfo) {
      if (!syncLocaleFiles.has(locale) && !asyncLocaleFiles.has(locale)) {
        ;(lazy ? asyncLocaleFiles : syncLocaleFiles).add(locale)
      }
    }
  }

  const importMapper = new Map<string, string>()
  for (const { code, path } of syncLocaleFiles) {
    importMapper.set(code, templateUtils.importName(`locale_${code}`))
    genCode += `${genImport(path, templateUtils.importName(`locale_${code}`))}\n`
  }

  // prettier-ignore
  genCode += `${Object.entries(options).map(([rootKey, rootValue]) => {
    if (rootKey === 'nuxtI18nOptions') {
      let genCodes = `export const resolveNuxtI18nOptions = async (context) => {\n`
      genCodes += `  const ${rootKey} = Object({})\n`
      for (const [key, value] of Object.entries(rootValue)) {
        if (key === 'vueI18n') {
          const optionLoaderVariable = `${key}OptionsLoader`
          genCodes += `  const ${optionLoaderVariable} = ${isObject(value)
            ? `async (context) => ${toCode(value)}\n`
            : isString(value)
              ? `async (context) => import(${toCode(value)}).then(r => (r.default || r)(context))\n`
              : `async (context) => ${toCode({})}\n`
          }`
          genCodes += `  ${rootKey}.${key} = await ${optionLoaderVariable}(context)\n`
        } else {
          genCodes += `  ${rootKey}.${key} = ${toCode(value)}\n`
        }
      }
      genCodes += `  return nuxtI18nOptions\n`
      genCodes += `}\n`
      return genCodes
    } else if (rootKey === 'nuxtI18nOptionsDefault') {
      // generate default nuxtI18n options
      return `export const ${rootKey} = Object({${Object.entries(rootValue).map(([key, value]) => {
        return `${key}: ${toCode(value)}`
      }).join(`,`)}})\n`
    } else if (rootKey === 'nuxtI18nInternalOptions') {
      return `export const ${rootKey} = Object({${Object.entries(rootValue).map(([key, value]) => {
        return `${key}: ${toCode(value)}`
      }).join(`,`)}})\n`
    } else if (rootKey === 'localeInfo') {
      let codes = `export const loadMessages = async () => {\n`
      codes += `  const messages = Object({})\n`
      if (langDir) {
        for (const { code } of syncLocaleFiles) {
          codes += `  messages[${toCode(code)}] = ${importMapper.get(code)}\n`
        }
        for (const { code, path } of asyncLocaleFiles) {
          codes += `  messages[${toCode(code)}] = await import(${toCode(path)} /* webpackChunkName: ${toCode(path)} */).then(r => r.default || r)\n`
        }
      }
      codes += `  return Promise.resolve(messages)\n`
      codes += `}\n`
      return codes
	  } else {
	    return `export const ${rootKey} = ${toCode(rootValue)}\n`
	  }
  }).join('\n')}`

  debug('generate code', genCode)
  return genCode
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function stringifyObj(obj: Record<string, any>): string {
  return `Object({${Object.entries(obj)
    .map(([key, value]) => `${JSON.stringify(key)}:${toCode(value)}`)
    .join(`,`)}})`
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function toCode(code: any): string {
  if (code === null) {
    return `null`
  }

  if (code === undefined) {
    return `undefined`
  }

  if (isString(code)) {
    return JSON.stringify(code)
  }

  if (isRegExp(code) && code.toString) {
    return code.toString()
  }

  if (isFunction(code) && code.toString) {
    return `(${code.toString()})`
  }

  if (isArray(code)) {
    return `[${code.map(c => toCode(c)).join(`,`)}]`
  }

  if (isObject(code)) {
    return stringifyObj(code)
  }

  return code + ``
}
