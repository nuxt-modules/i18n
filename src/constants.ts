export const NUXT_I18N_MODULE_ID = '@nuxtjs/i18n'
export const VUE_I18N_PKG = 'vue-i18n'
export const SHARED_PKG = '@intlify/shared'
export const MESSAGE_COMPILER_PKG = '@intlify/message-compiler'
export const CORE_PKG = '@intlify/core'
export const CORE_BASE_PKG = '@intlify/core-base'
export const H3_PKG = '@intlify/h3'
export const UTILS_PKG = '@intlify/utils'
export const UTILS_H3_PKG = '@intlify/utils/h3'
export const UFO_PKG = 'ufo'
export const IS_HTTPS_PKG = 'is-https'

// Options
const STRATEGY_PREFIX = 'prefix'
const STRATEGY_PREFIX_EXCEPT_DEFAULT = 'prefix_except_default'
const STRATEGY_PREFIX_AND_DEFAULT = 'prefix_and_default'
const STRATEGY_NO_PREFIX = 'no_prefix'
export const STRATEGIES = {
  PREFIX: STRATEGY_PREFIX,
  PREFIX_EXCEPT_DEFAULT: STRATEGY_PREFIX_EXCEPT_DEFAULT,
  PREFIX_AND_DEFAULT: STRATEGY_PREFIX_AND_DEFAULT,
  NO_PREFIX: STRATEGY_NO_PREFIX
} as const

export const DEFAULT_DYNAMIC_PARAMS_KEY = 'nuxtI18n'
export const DEFAULT_COOKIE_KEY = 'i18n_redirected'
export const SWITCH_LOCALE_PATH_LINK_IDENTIFIER = 'nuxt-i18n-slp'

export const DEFAULT_OPTIONS = {
  experimental: {
    localeDetector: '',
    switchLocalePathLinkSSR: false,
    autoImportTranslationFunctions: false
  },
  bundle: {
    compositionOnly: true,
    runtimeOnly: false,
    fullInstall: true,
    dropMessageCompiler: false
  },
  compilation: {
    jit: true,
    strictMessage: true,
    escapeHtml: false
  },
  customBlocks: {
    defaultSFCLang: 'json',
    globalSFCScope: false
  },
  vueI18n: '',
  locales: [] as string[],
  defaultLocale: '',
  defaultDirection: 'ltr',
  routesNameSeparator: '___',
  trailingSlash: false,
  defaultLocaleRouteNameSuffix: 'default',
  strategy: STRATEGY_PREFIX_EXCEPT_DEFAULT,
  lazy: false,
  langDir: null,
  rootRedirect: undefined,
  detectBrowserLanguage: {
    alwaysRedirect: false,
    cookieCrossOrigin: false,
    cookieDomain: null,
    cookieKey: DEFAULT_COOKIE_KEY,
    cookieSecure: false,
    fallbackLocale: '',
    redirectOn: 'root',
    useCookie: true
  },
  differentDomains: false,
  baseUrl: '',
  dynamicRouteParams: false,
  customRoutes: 'page',
  pages: {},
  skipSettingLocaleOnNavigate: false,
  types: 'composition',
  debug: false,
  parallelPlugin: false
} as const

export const NUXT_I18N_TEMPLATE_OPTIONS_KEY = 'i18n.options.mjs'
export const NUXT_I18N_COMPOSABLE_DEFINE_ROUTE = 'defineI18nRoute'
export const NUXT_I18N_COMPOSABLE_DEFINE_LOCALE = 'defineI18nLocale'
export const NUXT_I18N_COMPOSABLE_DEFINE_CONFIG = 'defineI18nConfig'
export const NUXT_I18N_COMPOSABLE_DEFINE_LOCALE_DETECTOR = 'defineI18nLocaleDetector'

export const TS_EXTENSIONS = ['.ts', '.cts', '.mts']
export const JS_EXTENSIONS = ['.js', '.cjs', '.mjs']
export const EXECUTABLE_EXTENSIONS = [...JS_EXTENSIONS, ...TS_EXTENSIONS]

export const NULL_HASH = '00000000'
