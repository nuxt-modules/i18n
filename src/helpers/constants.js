/**
 * @typedef {import('../../types').Options} Options
 */

// Internals
export const ROOT_DIR = 'nuxt-i18n'

// Options
const STRATEGY_PREFIX = 'prefix'
const STRATEGY_PREFIX_EXCEPT_DEFAULT = 'prefix_except_default'
const STRATEGY_PREFIX_AND_DEFAULT = 'prefix_and_default'
const STRATEGY_NO_PREFIX = 'no_prefix'
export const STRATEGIES = {
  PREFIX: STRATEGY_PREFIX,
  PREFIX_EXCEPT_DEFAULT: STRATEGY_PREFIX_EXCEPT_DEFAULT,
  PREFIX_AND_DEFAULT: STRATEGY_PREFIX_AND_DEFAULT,
  NO_PREFIX: STRATEGY_NO_PREFIX
}

export const COMPONENT_OPTIONS_KEY = 'nuxtI18n'

/** @type {Options} */
export const DEFAULT_OPTIONS = {
  vueI18n: {},
  vueI18nLoader: false,
  locales: [],
  defaultLocale: '',
  defaultDirection: 'ltr',
  routesNameSeparator: '___',
  defaultLocaleRouteNameSuffix: 'default',
  sortRoutes: true,
  strategy: STRATEGY_PREFIX_EXCEPT_DEFAULT,
  lazy: false,
  langDir: null,
  rootRedirect: null,
  detectBrowserLanguage: {
    alwaysRedirect: false,
    cookieCrossOrigin: false,
    cookieDomain: null,
    cookieKey: 'i18n_redirected',
    cookieSecure: false,
    fallbackLocale: '',
    onlyOnNoPrefix: false,
    onlyOnRoot: false,
    useCookie: true
  },
  differentDomains: false,
  seo: false,
  baseUrl: '',
  vuex: {
    moduleName: 'i18n',
    syncLocale: false,
    syncMessages: false,
    syncRouteParams: true
  },
  parsePages: true,
  pages: {},
  skipSettingLocaleOnNavigate: false,
  beforeLanguageSwitch: () => null,
  onBeforeLanguageSwitch: () => {},
  onLanguageSwitched: () => null
}
