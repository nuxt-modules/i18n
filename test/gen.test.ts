import { parse } from '@babel/parser'
import { DEFAULT_OPTIONS } from '../src/constants'
import { generateLoaderOptions } from '../src/gen'

import type { NuxtI18nOptions, NuxtI18nInternalOptions, VueI18nConfigPathInfo } from '../src/types'

vi.mock('node:fs')

beforeEach(async () => {
  vi.spyOn(await import('node:fs'), 'readFileSync').mockReturnValue('export default {}')
})

afterEach(() => {
  vi.clearAllMocks()
})

const LOCALE_CODES = ['en', 'ja', 'fr']
const LOCALE_INFO = [
  {
    code: 'en',
    file: 'en.json',
    path: '/path/to/en.json'
  },
  {
    code: 'ja',
    file: 'ja.json',
    path: '/path/to/ja.json'
  },
  {
    code: 'fr',
    file: 'fr.json',
    path: '/path/to/fr.json'
  }
]
const NUXT_I18N_OPTIONS = {
  defaultLocale: 'en'
} as NuxtI18nOptions

const NUXT_I18N_INTERNAL_OPTIONS = {
  __normalizedLocales: [
    {
      code: 'en'
    }
  ]
} as NuxtI18nInternalOptions

const NUXT_I18N_VUE_I18N_CONFIG = {
  absolute: '/path/to/i18n.config.ts',
  relative: 'i18n.config.ts',
  rootDir: '/path/to',
  relativeBase: '..'
} as VueI18nConfigPathInfo

function validateSyntax(code: string): boolean {
  let ret = false
  try {
    const node = parse(code, {
      allowImportExportEverywhere: true,
      sourceType: 'module',
      plugins: ['importAssertions']
    })
    ret = !node.errors.length
  } catch (e) {
    console.error(e)
  }
  return ret
}

test('basic', async () => {
  const { generateLoaderOptions } = await import('../src/gen')
  const code = generateLoaderOptions(
    false,
    '..',
    NUXT_I18N_VUE_I18N_CONFIG,
    [],
    {
      localeCodes: LOCALE_CODES,
      localeInfo: LOCALE_INFO,
      nuxtI18nOptions: NUXT_I18N_OPTIONS,
      nuxtI18nOptionsDefault: DEFAULT_OPTIONS,
      nuxtI18nInternalOptions: NUXT_I18N_INTERNAL_OPTIONS
    },
    { ssg: false, dev: true, parallelPlugin: false }
  )
  expect(validateSyntax(code)).toBe(true)
  expect(code).toMatchSnapshot()
})

test('lazy', () => {
  const code = generateLoaderOptions(
    true,
    '..',
    NUXT_I18N_VUE_I18N_CONFIG,
    [],
    {
      localeCodes: LOCALE_CODES,
      localeInfo: LOCALE_INFO,
      nuxtI18nOptions: NUXT_I18N_OPTIONS,
      nuxtI18nInternalOptions: NUXT_I18N_INTERNAL_OPTIONS
    },
    { ssg: false, dev: true, parallelPlugin: false }
  )
  expect(validateSyntax(code)).toBe(true)
  expect(code).toMatchSnapshot()
})

test('multiple files', () => {
  const code = generateLoaderOptions(
    true,
    '..',
    NUXT_I18N_VUE_I18N_CONFIG,
    [],
    {
      localeCodes: [...LOCALE_CODES, 'es', 'es-AR'],
      localeInfo: [
        ...LOCALE_INFO,
        ...[
          {
            code: 'es',
            file: 'es.json',
            path: '/path/to/es.json'
          },
          {
            code: 'es-AR',
            files: ['es.json', 'es-AR.json'],
            paths: ['/path/to/es.json', '/path/to/es-AR.json']
          }
        ]
      ],
      nuxtI18nOptions: NUXT_I18N_OPTIONS,
      nuxtI18nInternalOptions: NUXT_I18N_INTERNAL_OPTIONS
    },
    { ssg: false, dev: true, parallelPlugin: false }
  )
  expect(validateSyntax(code)).toBe(true)
  expect(code).toMatchSnapshot()
})

test('locale file in nested', () => {
  const code = generateLoaderOptions(
    true,
    '..',
    NUXT_I18N_VUE_I18N_CONFIG,
    [],
    {
      localeCodes: LOCALE_CODES,
      localeInfo: [
        {
          code: 'en',
          file: 'en/main.json',
          path: '/path/to/en.json'
        },
        {
          code: 'ja',
          file: 'ja/main.json',
          path: '/path/to/ja.json'
        },
        {
          code: 'fr',
          file: 'fr/main.json',
          path: '/path/to/fr.json'
        }
      ],
      nuxtI18nOptions: NUXT_I18N_OPTIONS,
      nuxtI18nInternalOptions: NUXT_I18N_INTERNAL_OPTIONS
    },
    { ssg: false, dev: true, parallelPlugin: false }
  )
  expect(validateSyntax(code)).toBe(true)
  expect(code).toMatchSnapshot()
})

test('vueI18n option', () => {
  const code = generateLoaderOptions(
    false,
    '..',
    NUXT_I18N_VUE_I18N_CONFIG,
    [
      {
        absolute: '/path/layer1/i18n.custom.ts',
        relative: 'i18n.custom.ts',
        rootDir: '/path/layer1',
        relativeBase: '../..'
      },
      {
        absolute: '/path/foo/layer2/vue-i18n.options.js',
        relative: 'vue-i18n.options.js',
        rootDir: '/path/foo/layer2',
        relativeBase: '../../..'
      }
    ],
    {
      localeCodes: LOCALE_CODES,
      localeInfo: LOCALE_INFO,
      nuxtI18nOptions: {
        vueI18n: 'vue-i18n.config.ts'
      },
      nuxtI18nInternalOptions: NUXT_I18N_INTERNAL_OPTIONS
    },
    { ssg: false, dev: true, parallelPlugin: false }
  )
  expect(validateSyntax(code)).toBe(true)
  expect(code).toMatchSnapshot()
})

test('toCode: function (arrow)', () => {
  const code = generateLoaderOptions(
    false,
    '..',
    NUXT_I18N_VUE_I18N_CONFIG,
    [],
    {
      localeCodes: LOCALE_CODES,
      nuxtI18nOptions: {
        ...NUXT_I18N_OPTIONS,
        locales: LOCALE_INFO.map(locale => ({
          ...locale,
          testFunc: (prop: unknown) => {
            return `Hello ${prop}`
          }
        }))
      },
      nuxtI18nOptionsDefault: DEFAULT_OPTIONS,
      nuxtI18nInternalOptions: NUXT_I18N_INTERNAL_OPTIONS
    },
    { ssg: false, dev: true, parallelPlugin: false }
  )
  expect(validateSyntax(code)).toBe(true)
  expect(code).toMatchSnapshot()
})

test('toCode: function (named)', () => {
  const code = generateLoaderOptions(false, '..', NUXT_I18N_VUE_I18N_CONFIG, [], {
    localeCodes: LOCALE_CODES,
    nuxtI18nOptions: {
      ...NUXT_I18N_OPTIONS,
      locales: LOCALE_INFO.map(locale => ({
        ...locale,
        testFunc(prop: unknown) {
          return `Hello ${prop}`
        }
      }))
    },
    nuxtI18nOptionsDefault: DEFAULT_OPTIONS,
    nuxtI18nInternalOptions: NUXT_I18N_INTERNAL_OPTIONS
  })
  expect(validateSyntax(code)).toBe(true)
  expect(code).toMatchSnapshot()
})
