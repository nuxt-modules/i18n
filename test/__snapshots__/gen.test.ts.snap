// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
"import locale_en_json_en from \\"../en.json\\" assert { type: \\"json\\" };
import locale_ja_json_ja from \\"../ja.json\\" assert { type: \\"json\\" };
import locale_fr_json_fr from \\"../fr.json\\" assert { type: \\"json\\" };
export const localeCodes = [\\"en\\",\\"ja\\",\\"fr\\"]

export const localeMessages = {
  \\"en\\": [{ key: \\"../en.json\\", load: () => Promise.resolve(locale_en_json_en) }],
  \\"ja\\": [{ key: \\"../ja.json\\", load: () => Promise.resolve(locale_ja_json_ja) }],
  \\"fr\\": [{ key: \\"../fr.json\\", load: () => Promise.resolve(locale_fr_json_fr) }],
}

export const resolveNuxtI18nOptions = async (context) => {
  const nuxtI18nOptions = Object({})
  nuxtI18nOptions.defaultLocale = \\"en\\"
  return nuxtI18nOptions
}

export const nuxtI18nOptionsDefault = Object({experimental: Object({\\"jsTsFormatResource\\":false}),compilation: Object({\\"jit\\":true,\\"strictMessage\\":true,\\"escapeHtml\\":false}),vueI18n: \\"\\",locales: [],defaultLocale: \\"\\",defaultDirection: \\"ltr\\",routesNameSeparator: \\"___\\",trailingSlash: false,defaultLocaleRouteNameSuffix: \\"default\\",strategy: \\"prefix_except_default\\",lazy: false,langDir: null,rootRedirect: null,detectBrowserLanguage: Object({\\"alwaysRedirect\\":false,\\"cookieCrossOrigin\\":false,\\"cookieDomain\\":null,\\"cookieKey\\":\\"i18n_redirected\\",\\"cookieSecure\\":false,\\"fallbackLocale\\":\\"\\",\\"redirectOn\\":\\"root\\",\\"useCookie\\":true}),differentDomains: false,baseUrl: \\"\\",dynamicRouteParams: false,customRoutes: \\"page\\",pages: Object({}),skipSettingLocaleOnNavigate: false,types: \\"composition\\",debug: false,parallelPlugin: false})

export const nuxtI18nInternalOptions = Object({__normalizedLocales: [Object({\\"code\\":\\"en\\"})]})
export const NUXT_I18N_MODULE_ID = \\"@nuxtjs/i18n\\"
export const isSSG = false
export const parallelPlugin = false
"
`;

exports[`lazy 1`] = `
"export const localeCodes = [\\"en\\",\\"ja\\",\\"fr\\"]

export const localeMessages = {
  \\"en\\": [{ key: \\"../en.json\\", load: () => import(\\"../en.json\\" /* webpackChunkName: \\"lang_en_json_en_json\\" */) }],
  \\"ja\\": [{ key: \\"../ja.json\\", load: () => import(\\"../ja.json\\" /* webpackChunkName: \\"lang_ja_json_ja_json\\" */) }],
  \\"fr\\": [{ key: \\"../fr.json\\", load: () => import(\\"../fr.json\\" /* webpackChunkName: \\"lang_fr_json_fr_json\\" */) }],
}

export const resolveNuxtI18nOptions = async (context) => {
  const nuxtI18nOptions = Object({})
  nuxtI18nOptions.defaultLocale = \\"en\\"
  return nuxtI18nOptions
}

export const nuxtI18nInternalOptions = Object({__normalizedLocales: [Object({\\"code\\":\\"en\\"})]})
export const NUXT_I18N_MODULE_ID = \\"@nuxtjs/i18n\\"
export const isSSG = false
export const parallelPlugin = false
"
`;

exports[`locale file in nested 1`] = `
"export const localeCodes = [\\"en\\",\\"ja\\",\\"fr\\"]

export const localeMessages = {
  \\"en\\": [{ key: \\"../en/main.json\\", load: () => import(\\"../en/main.json\\" /* webpackChunkName: \\"lang_en_en_main_json\\" */) }],
  \\"ja\\": [{ key: \\"../ja/main.json\\", load: () => import(\\"../ja/main.json\\" /* webpackChunkName: \\"lang_ja_ja_main_json\\" */) }],
  \\"fr\\": [{ key: \\"../fr/main.json\\", load: () => import(\\"../fr/main.json\\" /* webpackChunkName: \\"lang_fr_fr_main_json\\" */) }],
}

export const resolveNuxtI18nOptions = async (context) => {
  const nuxtI18nOptions = Object({})
  nuxtI18nOptions.defaultLocale = \\"en\\"
  return nuxtI18nOptions
}

export const nuxtI18nInternalOptions = Object({__normalizedLocales: [Object({\\"code\\":\\"en\\"})]})
export const NUXT_I18N_MODULE_ID = \\"@nuxtjs/i18n\\"
export const isSSG = false
export const parallelPlugin = false
"
`;

exports[`multiple files 1`] = `
"export const localeCodes = [\\"en\\",\\"ja\\",\\"fr\\",\\"es\\",\\"es-AR\\"]

export const localeMessages = {
  \\"en\\": [{ key: \\"../en.json\\", load: () => import(\\"../en.json\\" /* webpackChunkName: \\"lang_en_json_en_json\\" */) }],
  \\"ja\\": [{ key: \\"../ja.json\\", load: () => import(\\"../ja.json\\" /* webpackChunkName: \\"lang_ja_json_ja_json\\" */) }],
  \\"fr\\": [{ key: \\"../fr.json\\", load: () => import(\\"../fr.json\\" /* webpackChunkName: \\"lang_fr_json_fr_json\\" */) }],
  \\"es\\": [{ key: \\"../es.json\\", load: () => import(\\"../es.json\\" /* webpackChunkName: \\"lang_es_json_es_json\\" */) }],
  \\"es-AR\\": [{ key: \\"../es.json\\", load: () => import(\\"../es.json\\" /* webpackChunkName: \\"lang_es_json_es_json\\" */) },{ key: \\"../es-AR.json\\", load: () => import(\\"../es-AR.json\\" /* webpackChunkName: \\"lang_es_AR_json_es_AR_json\\" */) }],
}

export const resolveNuxtI18nOptions = async (context) => {
  const nuxtI18nOptions = Object({})
  nuxtI18nOptions.defaultLocale = \\"en\\"
  return nuxtI18nOptions
}

export const nuxtI18nInternalOptions = Object({__normalizedLocales: [Object({\\"code\\":\\"en\\"})]})
export const NUXT_I18N_MODULE_ID = \\"@nuxtjs/i18n\\"
export const isSSG = false
export const parallelPlugin = false
"
`;

exports[`toCode: function (arrow) 1`] = `
"export const localeCodes = [\\"en\\",\\"ja\\",\\"fr\\"]

export const resolveNuxtI18nOptions = async (context) => {
  const nuxtI18nOptions = Object({})
  nuxtI18nOptions.defaultLocale = \\"en\\"
  nuxtI18nOptions.locales = [Object({\\"code\\":\\"en\\",\\"file\\":\\"en.json\\",\\"testFunc\\":((prop) => {
            return \`Hello \${prop}\`;
          })}),Object({\\"code\\":\\"ja\\",\\"file\\":\\"ja.json\\",\\"testFunc\\":((prop) => {
            return \`Hello \${prop}\`;
          })}),Object({\\"code\\":\\"fr\\",\\"file\\":\\"fr.json\\",\\"testFunc\\":((prop) => {
            return \`Hello \${prop}\`;
          })})]
  return nuxtI18nOptions
}

export const nuxtI18nOptionsDefault = Object({experimental: Object({\\"jsTsFormatResource\\":false}),compilation: Object({\\"jit\\":true,\\"strictMessage\\":true,\\"escapeHtml\\":false}),vueI18n: \\"\\",locales: [],defaultLocale: \\"\\",defaultDirection: \\"ltr\\",routesNameSeparator: \\"___\\",trailingSlash: false,defaultLocaleRouteNameSuffix: \\"default\\",strategy: \\"prefix_except_default\\",lazy: false,langDir: null,rootRedirect: null,detectBrowserLanguage: Object({\\"alwaysRedirect\\":false,\\"cookieCrossOrigin\\":false,\\"cookieDomain\\":null,\\"cookieKey\\":\\"i18n_redirected\\",\\"cookieSecure\\":false,\\"fallbackLocale\\":\\"\\",\\"redirectOn\\":\\"root\\",\\"useCookie\\":true}),differentDomains: false,baseUrl: \\"\\",dynamicRouteParams: false,customRoutes: \\"page\\",pages: Object({}),skipSettingLocaleOnNavigate: false,types: \\"composition\\",debug: false,parallelPlugin: false})

export const nuxtI18nInternalOptions = Object({__normalizedLocales: [Object({\\"code\\":\\"en\\"})]})
export const NUXT_I18N_MODULE_ID = \\"@nuxtjs/i18n\\"
export const isSSG = false
export const parallelPlugin = false
"
`;

exports[`toCode: function (named) 1`] = `
"export const localeCodes = [\\"en\\",\\"ja\\",\\"fr\\"]

export const resolveNuxtI18nOptions = async (context) => {
  const nuxtI18nOptions = Object({})
  nuxtI18nOptions.defaultLocale = \\"en\\"
  nuxtI18nOptions.locales = [Object({\\"code\\":\\"en\\",\\"file\\":\\"en.json\\",\\"testFunc\\":(function (prop) {
          return \`Hello \${prop}\`;
        })}),Object({\\"code\\":\\"ja\\",\\"file\\":\\"ja.json\\",\\"testFunc\\":(function (prop) {
          return \`Hello \${prop}\`;
        })}),Object({\\"code\\":\\"fr\\",\\"file\\":\\"fr.json\\",\\"testFunc\\":(function (prop) {
          return \`Hello \${prop}\`;
        })})]
  return nuxtI18nOptions
}

export const nuxtI18nOptionsDefault = Object({experimental: Object({\\"jsTsFormatResource\\":false}),compilation: Object({\\"jit\\":true,\\"strictMessage\\":true,\\"escapeHtml\\":false}),vueI18n: \\"\\",locales: [],defaultLocale: \\"\\",defaultDirection: \\"ltr\\",routesNameSeparator: \\"___\\",trailingSlash: false,defaultLocaleRouteNameSuffix: \\"default\\",strategy: \\"prefix_except_default\\",lazy: false,langDir: null,rootRedirect: null,detectBrowserLanguage: Object({\\"alwaysRedirect\\":false,\\"cookieCrossOrigin\\":false,\\"cookieDomain\\":null,\\"cookieKey\\":\\"i18n_redirected\\",\\"cookieSecure\\":false,\\"fallbackLocale\\":\\"\\",\\"redirectOn\\":\\"root\\",\\"useCookie\\":true}),differentDomains: false,baseUrl: \\"\\",dynamicRouteParams: false,customRoutes: \\"page\\",pages: Object({}),skipSettingLocaleOnNavigate: false,types: \\"composition\\",debug: false,parallelPlugin: false})

export const nuxtI18nInternalOptions = Object({__normalizedLocales: [Object({\\"code\\":\\"en\\"})]})
export const NUXT_I18N_MODULE_ID = \\"@nuxtjs/i18n\\"
export const isSSG = false
export const parallelPlugin = false
"
`;

exports[`vueI18n option 1`] = `
"import locale_en_json_en from \\"../en.json\\" assert { type: \\"json\\" };
import locale_ja_json_ja from \\"../ja.json\\" assert { type: \\"json\\" };
import locale_fr_json_fr from \\"../fr.json\\" assert { type: \\"json\\" };
export const localeCodes = [\\"en\\",\\"ja\\",\\"fr\\"]

export const localeMessages = {
  \\"en\\": [{ key: \\"../en.json\\", load: () => Promise.resolve(locale_en_json_en) }],
  \\"ja\\": [{ key: \\"../ja.json\\", load: () => Promise.resolve(locale_ja_json_ja) }],
  \\"fr\\": [{ key: \\"../fr.json\\", load: () => Promise.resolve(locale_fr_json_fr) }],
}

export const resolveNuxtI18nOptions = async (context) => {
  const nuxtI18nOptions = Object({})
 const vueI18nConfigLoader = async (loader) => {
            const config = await loader().then(r => r.default || r)
            return typeof config === 'object'
              ? config
              : typeof config === 'function'
                ? await config()
                : {}
          }
  nuxtI18nOptions.vueI18n = Object({})
  nuxtI18nOptions.vueI18n.messages ??= {}
  const deepCopy = (src, des, predicate) => {
            for (const key in src) {
              if (typeof src[key] === 'object') {
                if (!typeof des[key] === 'object') des[key] = {}
                deepCopy(src[key], des[key], predicate)
              } else {
                if (predicate) {
                  if (predicate(src[key], des[key])) {
                    des[key] = src[key]
                  }
                } else {
                  des[key] = src[key]
                }
              }
            }
          }
          const mergeMessages = async (messages, loader) => {
            const layerConfig = await vueI18nConfigLoader(loader)
            const vueI18n = layerConfig || {}
            const layerMessages = vueI18n.messages || {}
            for (const [locale, message] of Object.entries(layerMessages)) {
              messages[locale] ??= {}
              deepCopy(message, messages[locale])
            }
          }
  return nuxtI18nOptions
}

export const nuxtI18nInternalOptions = Object({__normalizedLocales: [Object({\\"code\\":\\"en\\"})]})
export const NUXT_I18N_MODULE_ID = \\"@nuxtjs/i18n\\"
export const isSSG = false
export const parallelPlugin = false
"
`;
